src/main.rs

```diff
@@ -147,7 +147,8 @@ fn create_app<'a, 'b>(
             Arg::with_name("rng")
                 .long("rng")
                 .help(
-                    "Random number generator parameters \"src=<entropy_source_path>,iommu=on|off\"",
+                    "Random number generator parameters \
+                    \"src=<entropy_source_path>,iommu=on|off,id=<device_id>\"",
```

Comment:
I think we should present the user with the option of defining a device id only when it makes sense, i.e. when the user can hot-plug and unplug it. Since we already added that for our device hotplug implementation, I believe our UI should not be modified with this PR.

The sole reason why we need a device ID for e.g. the RNG device here is for our internal restoring implementation and thus this should be kept internal. Same for our vsocks devices.

That means there's a part of our VmConfig that we're not exposing to the user and I believe it is just fine. In case there's a disagreement about that and we want all our config items to be visible and modifiable from the UI, then we need to have some sort of config overlay (let's call it `VmDefinition` for the sake of that discussion).for adding our internal data to the user visible configuration. And we'd snapshot that overlay and restore from it, not from the VM config. This would clearly and cleanly separate the VM config defined and modified by the user from what we internally add to it, but otoh would add some complexity as e.g. we'd need to create and restore from the `VmDefinition` instead of what we do today from `VmConfig`. Since `VmDefinition` would be an overlay on top of `VmConfig`, it would be fairly easy to build the former from the latter and amend it as e.g. we add devices that don't have a user-visible/config-exposed id.

Bottom line: For now I'm personally ok if we overload our `VmConfig` with user invisible data, but we can think about a `VmConfig` overlay for a cleaner implementation in the future.

Reply:
The point of having an `id` field in the `VmConfig` is to rely on it to find out if there's a corresponding entry in the device tree or not, which means we can restore from it, or simply creating the device from scratch.
The only devices for which we can accept no `id` field in the `VmConfig` are the single devices like `rng`, `console` and `mem`.
Since we can create multiple `vsock` devices, we need to store the information in the `VmConfig` so that we can do this comparison with the device tree during the restore path.

The overlay layer might be an option by making it part of the `DeviceManagerState`. That would store an extended version of the `VmConfig`, but this is gonna grow our complexity.

About the unplug, it's taken care of as part of this PR, as I find the virtio device type associated with the id, and I return an error if the type is not expected to be removed.

Reply:
> The point of having an `id` field in the `VmConfig` is to rely on it to find out if there's a corresponding entry in the device tree or not, which means we can restore from it, or simply creating the device from scratch.

I'm certainly not arguing against the idea of having each device uniquely identified to be able to restore from it.

> The only devices for which we can accept no `id` field in the `VmConfig` are the single devices like `rng`, `console` and `mem`.

Could we remove them for those devices?

> Since we can create multiple `vsock` devices, we need to store the information in the `VmConfig` so that we can do this comparison with the device tree during the restore path.
>
>
> The overlay layer might be an option by making it part of the `DeviceManagerState`. That would store an extended version of the `VmConfig`, but this is gonna grow our complexity.
>

Then we should at least not allow the user to set the `vsock` device id.
 
> About the unplug, it's taken care of as part of this PR, as I find the virtio device type associated with the id, and I return an error if the type is not expected to be removed.

Yes, and that's fine for now. But the fact that we can potentially get a request to unplug a vsock device means we could improve this design in the future, by keeping ids for static devices completely unreachable to the user.

Reply:
I don't think we should let the user specify an id in this case. For "internal" ids we should prefix them with an `_` and never let the user specify anything with an `_` so that we know the use can't specify one that would clash with our "internal" ones.